### PR TITLE
Fix downloading agent manifest from upstream

### PR DESCRIPTION
**Pull request by MichaelKatsoulis**
This PR tries to fix bug created with https://github.com/elastic/elastic-package/pull/452

- Closes Issue: https://github.com/elastic/elastic-package/issues/458

In this PR a retry operation is added when downloading elastic agent manifest from upstream. If the status code of the response is not 200 or the received bytes of the file are less than 2000, the code will retry for 5 times to fetch the file before failing.
This way short network issues can be faced

### CODE REVIEW

File: internal/testrunner/runners/system/servicedeployer/kubernetes.go [hunk 1]
```diff
@@ -144,14 +151,84 @@ func findKubernetesDefinitions(definitionsDir string) ([]string, error) {
...
+func downloadElasticAgentManagedYAML(url string) ([]byte, error) {
+	// Get the data
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get file from URL %s", url)
+	}
+	defer resp.Body.Close()
+	logger.Debugf("status code when downloading elastic-agent-managed-kubernetes.yaml is %d", resp.StatusCode)
+	if resp.StatusCode != 200 {
+		return nil, errors.Wrapf(err, "downloading failed due to status code %d", resp.StatusCode)
```
Comment (mtojek) [inline internal/testrunner/runners/system/servicedeployer/kubernetes.go:176]: Maybe print also the body here?

File: internal/testrunner/runners/system/servicedeployer/kubernetes.go [hunk 2]
```diff
@@ -144,14 +151,84 @@ func findKubernetesDefinitions(definitionsDir string) ([]string, error) {
...
+			time.Sleep(sleep)
+		}
+		elasticAgentManagedYaml, err := f(url)
+		if err == nil {
+			logger.Debugf("downloaded %d bytes", len(elasticAgentManagedYaml))
+			if len(elasticAgentManagedYaml) > 2000 {
+				return elasticAgentManagedYaml, nil
+			}
+			err = fmt.Errorf("bytes downloaded should be more than 2000 but where: %d", len(elasticAgentManagedYaml))
+			logger.Debugf("failed because of %s", err)
```
Comment (mtojek) [inline internal/testrunner/runners/system/servicedeployer/kubernetes.go:229]: Did you check these debugf calls (for example with an invalid URL)?

File: internal/testrunner/runners/system/servicedeployer/kubernetes.go [hunk 3]
```diff
@@ -144,14 +151,84 @@ func findKubernetesDefinitions(definitionsDir string) ([]string, error) {
...
+			logger.Debugf("file downloaded is %s", string(elasticAgentManagedYaml))
```
Comment (mtojek) [inline internal/testrunner/runners/system/servicedeployer/kubernetes.go:230]: I discourage printing the doc here. If you really want to check if the content is right, maybe print just MD5?

File: internal/testrunner/runners/system/servicedeployer/kubernetes.go (hunk 2 above)
Comment (mtojek) [inline internal/testrunner/runners/system/servicedeployer/kubernetes.go:229]: I think you combine these two lines together

File: scripts/test-check-packages.sh [hunk 4]
```diff
@@ -10,7 +10,7 @@ cleanup() {
 
   # Dump kubectl details
   kubectl describe pods --all-namespaces > build/kubectl-dump.txt
-  kubectl logs -l app=kind-fleet-agent-clusterscope -n kube-system >> build/kubectl-dump.txt
```
Comment (mtojek) [inline scripts/test-check-packages.sh:13]: Please grep the codebase here if there are not references to `kind-fleet-agent-clusterscope `.

**Comment by MichaelKatsoulis** [comment]
I update the error handling and error messages. It is tested with scenarios where the url is invalid and the bytes are less than expected. The errors are the ones supposed to be.

File: scripts/test-check-packages.sh (hunk 4 above)
Comment (MichaelKatsoulis) [reply scripts/test-check-packages.sh:13]: There are only references in tests  and docs of kubernetes test package

**Review by mtojek — APPROVED** [review]

**Comment by mtojek** [comment]
This is unrelated to this PR, but maybe worth defining in the beats repo:

```
[2021-08-03T14:50:34.562Z] 2021/08/03 14:50:34 DEBUG downloading elastic-agent-managed-kubernetes.yaml from https://raw.githubusercontent.com/elastic/beats/7.x/deploy/kubernetes/elastic-agent-managed-kubernetes.yaml
[2021-08-03T14:50:34.827Z] 2021/08/03 14:50:34 DEBUG status code when downloading elastic-agent-managed-kubernetes.yaml is 200
[2021-08-03T14:50:34.827Z] 2021/08/03 14:50:34 DEBUG downloaded 5084 bytes
[2021-08-03T14:50:34.827Z] 2021/08/03 14:50:34 DEBUG Apply Kubernetes stdin
[2021-08-03T14:50:34.827Z] 2021/08/03 14:50:34 DEBUG run command: /var/lib/jenkins/workspace/t-manager_elastic-package_PR-461/bin/kubectl apply -f - -o yaml
[2021-08-03T14:50:35.399Z] 2021/08/03 14:50:35 DEBUG Handle "apply" command output
[2021-08-03T14:50:35.399Z] 2021/08/03 14:50:35 DEBUG Extract resources from command output
[2021-08-03T14:50:35.399Z] 2021/08/03 14:50:35 DEBUG Wait for ready resources
[2021-08-03T14:50:35.399Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent (kind: DaemonSet, namespace: kube-system)
[2021-08-03T14:50:35.399Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent (kind: ClusterRoleBinding, namespace: )
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent (kind: RoleBinding, namespace: kube-system)
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent-kubeadm-config (kind: RoleBinding, namespace: kube-system)
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent (kind: ClusterRole, namespace: )
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent (kind: Role, namespace: kube-system)
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent-kubeadm-config (kind: Role, namespace: kube-system)
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG Sync resource info: elastic-agent (kind: ServiceAccount, namespace: kube-system)
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG beginning wait for 8 resources with timeout of 10m0s
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG install custom Kubernetes definitions (directory: /var/lib/jenkins/workspace/t-manager_elastic-package_PR-461/src/github.com/elastic/elastic-package/test/packages/kubernetes/data_stream/apiserver/_dev/deploy/k8s)
[2021-08-03T14:50:35.659Z] 2021/08/03 14:50:35 DEBUG no custom definitions found (directory: /var/lib/jenkins/workspace/t-manager_elastic-package_PR-461/src/github.com/elastic/elastic-package/test/packages/kubernetes/data_stream/apiserver/_dev/deploy/k8s). Nothing else will be installed.
[2021-08-03T14:50:35.660Z] 2021/08/03 14:50:35 DEBUG GET http://127.0.0.1:5601/api/fleet/agents
[2021-08-03T14:50:35.660Z] 2021/08/03 14:50:35 DEBUG filter agents using criteria: NamePrefix=kind-control-plane
[2021-08-03T14:50:35.660Z] 2021/08/03 14:50:35 DEBUG found 0 enrolled agent(s)
[2021-08-03T14:50:36.597Z] 2021/08/03 14:50:36 DEBUG GET http://127.0.0.1:5601/api/fleet/agents
[2021-08-03T14:50:36.597Z] 2021/08/03 14:50:36 DEBUG filter agents using criteria: NamePrefix=kind-control-plane
[2021-08-03T14:50:36.597Z] 2021/08/03 14:50:36 DEBUG found 0 enrolled agent(s)
[2021-08-03T14:50:37.552Z] 2021/08/03 14:50:37 DEBUG GET http://127.0.0.1:5601/api/fleet/agents
[2021-08-03T14:50:37.811Z] 2021/08/03 14:50:37 DEBUG filter agents using criteria: NamePrefix=kind-control-plane
[2021-08-03T14:50:37.811Z] 2021/08/03 14:50:37 DEBUG found 0 enrolled agent(s)
[2021-08-03T14:50:38.752Z] 2021/08/03 14:50:38 DEBUG GET http://127.0.0.1:5601/api/fleet/agents
[2021-08-03T14:50:38.752Z] 2021/08/03 14:50:38 DEBUG filter agents using criteria: NamePrefix=kind-control-plane
[2021-08-03T14:50:38.752Z] 2021/08/03 14:50:38 DEBUG found 0 enrolled agent(s)
```

As you can see, the code doesn't work for created resources as it does for kube-state-metrics:

```
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Handle "apply" command output
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Extract resources from command output
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Wait for ready resources
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: kube-state-metrics (kind: ClusterRoleBinding, namespace: )
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: kube-state-metrics (kind: ClusterRole, namespace: )
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: hello (kind: CronJob, namespace: default)
[2021-08-03T14:56:34.402Z] W0803 14:56:34.222531  100343 warnings.go:70] batch/v1beta1 CronJob is deprecated in v1.21+, unavailable in v1.25+; use batch/v1 CronJob
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: kube-state-metrics (kind: Deployment, namespace: kube-system)
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: hello (kind: Job, namespace: default)
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: task-pv-volume (kind: PersistentVolume, namespace: )
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: task-pv-claim (kind: PersistentVolumeClaim, namespace: default)
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: pods-high (kind: ResourceQuota, namespace: default)
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: kube-state-metrics (kind: ServiceAccount, namespace: kube-system)
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: kube-state-metrics (kind: Service, namespace: kube-system)
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Sync resource info: web (kind: StatefulSet, namespace: default)
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG beginning wait for 11 resources with timeout of 10m0s
[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
[2021-08-03T14:56:36.307Z] 2021/08/03 14:56:36 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
[2021-08-03T14:56:38.871Z] 2021/08/03 14:56:38 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
[2021-08-03T14:56:40.780Z] 2021/08/03 14:56:40 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
[2021-08-03T14:56:42.693Z] 2021/08/03 14:56:42 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
[2021-08-03T14:56:44.604Z] 2021/08/03 14:56:44 DEBUG GET http://127.0.0.1:5601/api/fleet/agents
[2021-08-03T14:56:44.604Z] 2021/08/03 14:56:44 DEBUG filter agents using criteria: NamePrefix=kind-control-plane
[2021-08-03T14:56:44.604Z] 2021/08/03 14:56:44 DEBUG found 1 enrolled agent(s)
[2021-08-03T14:56:44.604Z] 2021/08/03 14:56:44 DEBUG creating test policy...
[2021-08-03T14:56:44.604Z] 2021/08/03 14:56:44 DEBUG POST http://127.0.0.1:5601/api/fleet/agent_policies
```

I was wondering if this is something we can improve here.

**Comment by MichaelKatsoulis** [comment]
>[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG beginning wait for 11 resources with timeout of 10m0s
>[2021-08-03T14:56:34.402Z] 2021/08/03 14:56:34 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
>[2021-08-03T14:56:36.307Z] 2021/08/03 14:56:36 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
>[2021-08-03T14:56:38.871Z] 2021/08/03 14:56:38 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
>[2021-08-03T14:56:40.780Z] 2021/08/03 14:56:40 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready
>[2021-08-03T14:56:42.693Z] 2021/08/03 14:56:42 DEBUG Deployment is not ready: kube-system/kube-state-metrics. 0 out of 1 expected pods are ready

You mean that the part of checking the readiness of the pods is missing?

**Comment by mtojek** [comment]
Yes. As you see [here](https://github.com/elastic/elastic-package/blob/master/internal/kubectl/kubectl_apply.go#L112), we're using helm internals (same logic) to wait for resources. I'm wondering what's missing that it doesn't wait for it. Is it just lack of healthcheck or deployment?

**Review by mtojek — CHANGES_REQUESTED** [review]
To be honest it sounds like a blocker for this, as with remote YAML elastic-package doesn't wait for agent deployment.

**Review by mtojek — CHANGES_REQUESTED** [review]
Please remove also this [file](https://github.com/elastic/elastic-package/blob/master/internal/install/static_kubernetes_elastic_agent_yml.go). I understand that it won't be used anymore.

**Review by mtojek — APPROVED** [review]
LGTM. Please merge it if CI is happy.

**Pull request merged**